pkg/auth/service.go

```diff
@@ -1147,10 +1152,32 @@ func NewAPIAuthServiceWithClient(client ClientWithResponsesInterface, externalPr
 	}, nil
 }
 
-func CheckPermissions(ctx context.Context, node permissions.Node, username string, policies []*model.Policy) CheckResult {
+func (n *NeededPermissions) String() string {
+	w := &strings.Builder{}
+	if len(n.Denied) != 0 {
+		fmt.Fprintln(w, "denied from:")
+		for _, statement := range n.Denied {
+			for _, action := range statement.Action {
+				fmt.Fprintln(w, action)
+			}
+		}
+		return strings.TrimSpace(w.String())
+	}
+	if len(n.Unauthorized) != 0 {
+		fmt.Fprintln(w, "lacking permissions for:")
+		for _, node := range n.Unauthorized {
+			fmt.Fprintln(w, node.Permission.Action)
+		}
+		return strings.TrimSpace(w.String())
+	}
+	return strings.TrimSpace(w.String())
+}
+
+func CheckPermissions(ctx context.Context, node permissions.Node, username string, policies []*model.Policy, perm *NeededPermissions) CheckResult {
```

Comment:
nit: variable name `perm` is too generic. 
Can you use something more explicit? 
for example: `audit`, `denyAudit`, `permAudit`, `history` etc...
So that in the future when extending, debugging or just reading the code we can immediately understand what it is used for.